### PR TITLE
Tooltip redesign

### LINKED ISSUE #550: Single Tooltip Enhancement
Issue opened by Accurat to keep track of the discussions around the "Single Tooltip" design and development.

This is the last design iteration, still to be worked and discussed on:
image

### PR DESCRIPTION
**Pull request by theiliad**
Closes #550 

TODO:
- [x] Truncation
- [x] Stacked bar tooltip order fix

### CODE REVIEW

**Comment by natashadecoste** [comment]
do we care that the order of the tooltips isnt the order of the bars on a vertical stacked bar chart?
image

for horizontal it seems to make sense that the left -> right ordering of the bars does match top-> down ordering in the tooltip

**Comment by theiliad** [comment]
> do we care that the order of the tooltips isnt the order of the bars on a vertical stacked bar chart?
> image
> 
> for horizontal it seems to make sense that the left -> right ordering of the bars does match top-> down ordering in the tooltip

yeah the stacked bar order might have to change...

**Comment by natashadecoste** [comment]
sometimes when i hover on a dot and get the single datapoint tooltip, the dots all dissapear right after (but the tooltip stays). I think that all the dots should stay visible and hovered one filled

area demo: Area (time series - natural curve)

File: packages/core/src/components/graphs/scatter.ts [hunk 1]
```diff
@@ -318,21 +318,26 @@ export class Scatter extends Component {
...
+					data: overlappingData
 				});
 
-				const eventNameToDispatch =
-					d3Event.type === "mouseover"
-						? Events.Scatter.SCATTER_MOUSEOVER
-						: Events.Scatter.SCATTER_MOUSEMOVE;
 				// Dispatch mouse event
-				self.services.events.dispatchEvent(eventNameToDispatch, {
+				self.services.events.dispatchEvent(Events.Bar.BAR_MOUSEOVER, {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/scatter.ts:325]:
```suggestion
				self.services.events.dispatchEvent(Events.Scatter.SCATTER_MOUSEOVER, {
```

File: packages/core/src/components/graphs/scatter.ts [hunk 2]
```diff
@@ -318,21 +318,26 @@ export class Scatter extends Component {
...
 					element: hoveredElement,
 					datum
 				});
 			})
+			.on("mousemove", function (datum) {
+				const hoveredElement = select(this);
+
+				// Dispatch mouse event
+				self.services.events.dispatchEvent(Events.Bar.BAR_MOUSEMOVE, {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/scatter.ts:334]:
```suggestion
				self.services.events.dispatchEvent(Events.Scatter.SCATTER_MOUSEMOVE, {
```

File: packages/core/src/components/graphs/pie.ts [hunk 3]
```diff
@@ -13,6 +13,7 @@ import {
 import { select } from "d3-selection";
 import { arc, pie } from "d3-shape";
 import { interpolate } from "d3-interpolate";
+import { dragDisable } from "d3";
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/pie.ts:16]:
```suggestion
```

File: packages/core/src/components/essentials/tooltip-axis.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		const { cartesianScales } = this.services;
+		const domainAxisOptions = cartesianScales.getDomainAxisOptions();
+		const domainIdentifier = cartesianScales.getDomainIdentifier();
+		const domainAxisScaleType = cartesianScales.getDomainAxisScaleType();
+		const rangeAxisOptions = cartesianScales.getRangeAxisOptions();
+		const rangeIdentifier = cartesianScales.getRangeIdentifier();
+		const rangeAxisScaleType = cartesianScales.getRangeAxisScaleType();
+
+		// Generate default tooltip
+		const { groupMapsTo } = this.model.getOptions().data;
```
Comment (natashadecoste) [inline packages/core/src/components/essentials/tooltip-axis.ts:25]:
```suggestion
		const { groupMapsTo } = options.data;
```

**Comment by theiliad** [comment]
> sometimes when i hover on a dot and get the single datapoint tooltip, the dots all dissapear right after (but the tooltip stays). I think that all the dots should stay visible and hovered one filled
> 
> area demo: Area (time series - natural curve)

that's an area chart fix, probably a different PR

**Pull request merged**